src/data/sym/card.lean

```diff
@@ -36,9 +35,185 @@ Prove the general case of stars and bars.
+  { split_ifs,
+    { rw pos_iff_ne_zero,
+      rintro rfl,
+      obtain ⟨w, r⟩ := @multiset.exists_mem_of_ne_zero _ x.val (λ h, by simpa [h] using x.property),
+      simpa [subsingleton.elim w 0] using r, },
+    { cases lt_or_eq_of_le (nat.le_of_lt_succ a.property); solve_by_elim } },
+  { rw [multiset.card_map, x.property] }
+end
+
+def decode (n k : ℕ) : sym (fin n) k.succ ⊕ sym (fin n.succ) k → sym (fin n.succ) k.succ
+| (sum.inl x) := ⟨x.val.map (λ a, ⟨a.val, a.property.step⟩),
+                  by simpa [multiset.card_map] using x.property⟩
+| (sum.inr x) := ⟨multiset.cons (fin.last n) x.val, by rw [multiset.card_cons, x.property]⟩
+
+lemma equivalent (n k : ℕ) : sym (fin n.succ) k.succ ≃ sym (fin n) k.succ ⊕ sym (fin n.succ) k :=
+{ to_fun := encode n k,
+  inv_fun := decode n k,
+  left_inv := begin
+    rintro ⟨x, hx⟩,
+    rw encode,
+    split_ifs,
+    { rw decode,
+      simp [multiset.cons_erase h], },
+    { simp only [decode, multiset.map_map, subtype.mk.inj_eq, function.comp],
+      convert multiset.map_congr _,
+      { rw multiset.map_id, },
+      rintros ⟨g, hg⟩ h',
+      split_ifs with h'',
+      { simp only [fin.coe_mk] at h'',
+        subst g,
+        exact false.elim (h h'), },
+      { refl } },
+  end,
+  right_inv := begin
+    rintro (x|x),
+    { cases x with x hx,
+      rw [decode, encode],
+      split_ifs,
+      { obtain ⟨y, v, b⟩ := multiset.mem_map.mp h,
+        rw [fin.last, subtype.mk_eq_mk] at b,
+        have := y.property,
+        rw b at this,
+        exact nat.lt_asymm this this, },
+      { simp only [multiset.map_map, function.comp],
+        simp only [fin.val_eq_coe, subtype.mk_eq_mk, multiset.map_map, fin.coe_mk],
+        convert multiset.map_congr _,
+        { rw multiset.map_id, },
+        rintros ⟨g, hg⟩ h',
+        split_ifs with h'',
+        { simp only [fin.coe_mk] at h'',
+          subst g,
+          exact (nat.lt_asymm hg hg).elim, },
+        { refl } } },
+    { rw [decode, encode],
+      split_ifs,
+      { simp, },
+      { apply h,
+        apply multiset.mem_cons_self, } }
+  end }
+
+lemma multichoose1_rec (n k : ℕ) :
+  multichoose1 n.succ k.succ = multichoose1 n k.succ + multichoose1 n.succ k :=
+begin
+  simp only [multichoose1, fintype.card_sum.symm],
+  exact fintype.card_congr (equivalent n k),
+end
+
+lemma multichoose2_rec (n k : ℕ) :
+  multichoose2 n.succ k.succ = multichoose2 n k.succ + multichoose2 n.succ k :=
+begin
+  simp only [multichoose2, nat.add_succ, tsub_zero, nat.succ_sub_succ_eq_sub, nat.succ_add_sub_one,
+    nat.succ_add, nat.choose_succ_succ, nat.add_comm],
+end
+
+lemma multichoose1_eq_multichoose2 : ∀ (n k : ℕ), multichoose1 n k = multichoose2 n k
+| 0 0 := begin
+  simp only [multichoose1, multichoose2],
+  norm_num,
+  dec_trivial,
+end
+| 0 (k + 1) := begin
+  simp only [multichoose1, multichoose2],
+  norm_num,
+  have no_inhabitants : sym (fin 0) k.succ → false := begin
```

Comment:
This should be stated globally as:
```lean
instance [is_empty α] : is_empty (sym2 α n) :=
````

Reply:
You mean
```lean
instance [is_empty α] : is_empty (sym2 α) :=

instance [is_empty α] : is_empty (sym α n) :=
```
While you're at it, you can also provide
```lean
instance [unique α] : unique (sym2 α) :=

instance [unique α] : unique (sym α n) :=

instance [subsingleton α] : subsingleton (sym2 α) :=

instance [subsingleton α] : subsingleton (sym α n) :=
````

Reply:
`instance [is_empty α] : is_empty (sym α n)` can't be provided, `sym α 0` has one inhabitant if `α` is false. I will prove subsingleton for the n = 0 case and is_empty for the n.succ case.

Reply:
Nice catch, my mistake.

I don't think you need to special case n=0, as
```
instance [subsingleton α] : subsingleton (sym α n) :=
```
Should cover that